src/value/owned.rs

```diff
@@ -1,186 +1,183 @@
-//!
-//! This API is simpler to manage with respect to borrowing lifetimes, but requires full
-//! ownership of data to do so.
-
 use crate::ion_eq::IonEq;
 use crate::text::text_formatter::IonValueFormatter;
 use crate::types::decimal::Decimal;
 use crate::types::integer::Integer;
 use crate::types::timestamp::Timestamp;
-use crate::types::SymbolId;
 use crate::{IonType, Symbol};
 use num_bigint::BigInt;
 use std::collections::HashMap;
 use std::fmt::{Display, Formatter};
 use std::iter::FromIterator;
-use std::rc::Rc;
 
 use crate::symbol_ref::AsSymbolRef;
+use crate::value::builders::{ListBuilder, SExpBuilder, StructBuilder};
 use crate::value::iterators::{
     ElementsIterator, FieldIterator, FieldValuesIterator, IndexVec, SymbolsIterator,
 };
 
-impl Symbol {
-    pub fn symbol_id(&self) -> Option<SymbolId> {
-        match self.text() {
-            Some(_) => None,
-            None => Some(0),
-        }
+impl Element {
+    pub fn null(null_type: IonType) -> Element {
+        null_type.into()
     }
 
-    pub fn with_text(self, text: &'static str) -> Self {
-        Symbol::owned(text)
+    pub fn boolean(value: bool) -> Element {
+        value.into()
     }
 
-    pub fn with_symbol_id(self, _local_sid: SymbolId) -> Self {
-        // Because `Symbol` represents a fully resolved symbol...
-        if self.text().is_some() {
-            // ...if we already have text, we can discard the symbol ID.
-            return self;
-        }
-        // Otherwise, the text is unknown.
-        Symbol::unknown_text()
+    pub fn string<I: Into<String>>(str: I) -> Element {
+        let text: String = str.into();
+        text.into()
     }
 
-    pub fn text_token(text: &'static str) -> Self {
-        Symbol::owned(text)
+    pub fn symbol<I: Into<Symbol>>(symbol: I) -> Element {
+        let symbol: Symbol = symbol.into();
+        symbol.into()
     }
 
-    pub fn symbol_id_token(_local_sid: SymbolId) -> Self {
-        // Because `Symbol` represents a fully resolved symbol, constructing one from a symbol ID
-        // alone means that it has no defined text and is therefore equivalent to SID 0.
-        Symbol::unknown_text()
+    pub fn integer<I: Into<Integer>>(integer: I) -> Element {
+        let integer: Integer = integer.into();
+        integer.into()
     }
-}
 
-/// Constructs a [`Symbol`] with just text.
-/// A common case for text and synthesizing tokens.
-#[inline]
-pub fn text_token(text: &str) -> Symbol {
-    Symbol::owned(text)
-}
+    pub fn decimal(decimal: Decimal) -> Element {
+        decimal.into()
+    }
 
-/// An owned implementation of [`Builder`].
-impl Element {
-    fn new_null(e_type: IonType) -> Element {
-        Value::Null(e_type).into()
+    pub fn timestamp(timestamp: Timestamp) -> Element {
+        timestamp.into()
     }
 
-    fn new_bool(bool: bool) -> Element {
-        Value::Boolean(bool).into()
+    pub fn float(float: f64) -> Element {
```

Comment:
🗺️ I renamed the `new_f64` method to `new_float` for consistent use of Ion type names.

Reply:
Can a float be case into a specific float type, e.g. `Element.as_f64`?

Reply:
In the Ion data model, a `float` is always 64 bits. Once it's read from the stream into memory, it's basically always an `f64`. However, on the writing side we make an effort to encode it as an `f32` if it can be done losslessly.